scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingDSLCompilationService.java

```diff
@@ -0,0 +1,102 @@
+
+    try {
+      this.nodeProcess.getInputStream();
+      this.nodeProcess.getOutputStream();
+    } catch (Exception e) {
+      throw new SchedulingDSLCompilationException("Could not create node subprocess: ", e);
+    }
+  }
+
+  public void close() {
+    this.nodeProcess.destroy();
+  }
+
+  /**
+   * NOTE: This method is not re-entrant (assumes only one call to this method is running at any given time)
+   */
+  public SchedulingDSL.GoalSpecifier compileSchedulingGoalDSL(final PlanId planId, final String goalTypescript, final String goalName)
+  throws SchedulingDSLCompilationException, IOException
+  {
+    final var generatedCode = JSONObject.quote(this.typescriptCodeGenerationService.generateTypescriptTypesForPlan(planId));
+
+    /*
+    * PROTOCOL:
+    *   denote this java program as JAVA, and the node subprocess as NODE
+    *
+    *   JAVA -- stdin --> NODE: { "source": "sourcecode", "filename": "goalname", "generatedCode": "generatedcode" } \n
+    *   NODE -- stdout --> JAVA: one of "success\n" or "error\n"
+    *   NODE -- stdout --> JAVA: payload associated with success or failure, must be exactly one line terminated with \n
+    * */
+    final var inputWriter = this.nodeProcess.outputWriter();
+    final var outputReader = this.nodeProcess.inputReader();
+
+    final var quotedGoalTypescript = JSONObject.quote(goalTypescript); // adds extra quotes to start and end
+    inputWriter.write("{ \"source\": " + quotedGoalTypescript + ", \"filename\": \"" + goalName + "\", \"generatedCode\": " + generatedCode + " }\n");
+    inputWriter.flush();
+
+    final var status = outputReader.readLine();
+    if (status.equals("error")) {
+      throw new SchedulingDSLCompilationException(outputReader.readLine());
+    }
+
+    if (status.equals("success")) {
+      final var output = outputReader.readLine();
+      try {
+        return parseJson(output, SchedulingDSL.schedulingJsonP);
+      } catch (InvalidJsonException | InvalidEntityException e) {
+        throw new SchedulingDSLCompilationException("Could not parse JSON returned from typescript: ", e);
+      }
+    }
+
+    // Status was neither failure nor success, the protocol has been violated.
+    throw new Error("scheduling dsl compiler returned unexpected status: " + status);
+  }
+
+  private static <T> T parseJson(final String jsonStr, final JsonParser<T> parser) throws InvalidJsonException, InvalidEntityException {
+    try (final var reader = Json.createReader(new StringReader(jsonStr))) {
+      final var requestJson = reader.readValue();
+      final var result = parser.parse(requestJson);
+      return result.getSuccessOrThrow(reason -> new InvalidEntityException(List.of(reason)));
+    } catch (JsonParsingException e) {
+      throw new InvalidJsonException(e);
+    }
+  }
+
+  public static class SchedulingDSLCompilationException extends Exception {
+    SchedulingDSLCompilationException(final String message, final Exception e) {
```

Comment:
Initially the name of this exception made me think it was a true DSL compilation exception, not a DSL compilation *service* exception. Could maybe just make this `ServiceException` since it's a static class of `SchedulingDSLCompilationService` already. Not sure